### DISCUSSION TITLE
Indexing a large .jsonl that cannot be loaded in RAM

### DISCUSSION THREAD

**Discussion by liyucheng09** [xhluca/bm25s#29] [Q&A] [answered]
@xhluca Hi, thanks for the nice project.

I have a large .jsonl file where each line got a `text` and an `id`.
My question is, as I cannot load all lines in the RAM, i cannot obtain all ids before I start `.index()`.

But your implementation seems to require all ids for `retriever = bm25s.BM25(corpus=corpus_id)` before the indexing started.
Any ideas to handle this issue?

**Comment by xhluca** [comment, marked as answer]
The corpus you are passing to `BM25` is optional, it is only used during retrieval and when you are saving/loading the model. You can just initialize without the corpus:

```
retriever = bm25s.BM25()
```

That said, if you still want to pass the jsonl file to the retriever's initialization call, but don't have enough memory to load it in memory, you can take a look at the `bm25s.utils.corpus.JsonlCorpus` class which lets you read a jsonl dynamically through memory mapping: https://github.com/xhluca/bm25s/blob/73c7dea9ea7f88a23a7fa9a94e9a7bca48669f1c/bm25s/utils/corpus.py#L101-L184

Note however that before it does that, it needs to index the jsonl and create a `mmindex` file in the same directory as the jsonl file; after that, it will read any line of the jsonl extremely quickly, and even has basic support for slicing!

**Reply by ddofer** [reply]
Any tip for doing this over multiple jsonl files? (That add up to a large one)? (For doing indexing over them all)?

**Comment by liyucheng09** [comment]
@xhluca This is really cool. And thanks for your prompt response!

One quick question, I have tried to new a `BM25` without a given corpus, and I found `results, _ = retriever.retrieve` will just return the index of the document in the `corpus_tokens`.

I mean, it will use default ids assigned by the `BM25` if the corpus is not given for it? Is there a way to pass the id along side the tokenized corpus? like `retriever.index(corpus_tokens)` where `corpus_tokens` are a list of dict `[...{'id': ..., 'text': ...}...]`. So that the final result can be my customized ids.

**Reply by xhluca** [reply]
Yes, it will return the index when the corpus is not available. If you want it to return ids, you can use `retriever.retrieve(..., corpus=corpus_ids)`, or use the `JsonlCorpus` class I recommend above, e.g. try:

```python
corpus = JsonlCorpus('corpus_records.jsonl')
lines = corpus[indexes_returned_by_retriever]
```

The reason `index()` does not store the corpus (or any key in the corpus) is that it is unpredictable how much memory will be used the iterator; as an extension, storing certain keys but not others would add complexity to the `index` API (which already has many params) without clear benefit over using JsonlCorpus or passing a `corpus_ids` to `retriever.retrieve(corpus=corpus_ids)`.

If you still believe it is the best thing to do, feel free to subclass index:

```python
class CustomBM25(bm25s.BM25):
  def index(self, ...):
    # modify the index implementation here
```

**Reply by xhluca** [reply]
You could also read the jsonl yourself and create a cached ids file, i.e.:

```python
ids = []

with open('corpus.jsonl') as f:
  for line in f:
    record = json.load(line)
    ids.append(record['id'])

with open('corpus.ids.json', 'w') as f:
  json.dump(ids, f)
```
